File: rust/src/medium/grep.rs
use anyhow::Error;
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

#[derive(Debug, Default)]
pub struct Flags {
    line_numbers: bool,
    only_names: bool,
    case_insensitive: bool,
    invert: bool,
    entire_lines: bool,
}

impl Flags {
    pub fn new(flags: &[&str]) -> Self {
        flags
            .iter()
            .fold(Flags::default(), |mut flags, x| match *x {
                "-n" => {
                    flags.line_numbers = true;
                    flags
                }
                "-l" => {
                    flags.only_names = true;
                    flags
                }
                "-i" => {
                    flags.case_insensitive = true;
                    flags
                }
                "-v" => {
                    flags.invert = true;
                    flags
                }
                "-x" => {
                    flags.entire_lines = true;
                    flags
                }
                _ => flags,
            })
    }
}

fn is_match(line: &str, pattern: &str, flags: &Flags) -> bool {
    let is_match = match flags {
        Flags {
            entire_lines: true,
            case_insensitive: true,
            ..
        } => line.to_lowercase() == (pattern.to_lowercase()),
        Flags {
            entire_lines: true,
            case_insensitive: false,
            ..
        } => line == pattern,
        Flags {
            entire_lines: false,
            case_insensitive: true,
            ..
        } => line.to_lowercase().contains(&pattern.to_lowercase()),
        Flags {
            entire_lines: false,
            case_insensitive: false,
            ..
        } => line.contains(pattern),
    };

    if flags.invert { !is_match } else { is_match }
}

pub fn grep(pattern: &str, flags: &Flags, files: &[&str]) -> Result<Vec<String>, Error> {
    let sources: Result<Vec<_>, Error> = files
        .iter()
        .map(|file_name| {
            let file = File::open(file_name)?;
            Ok(BufReader::new(file)
                .lines()
                .flatten()
                .enumerate()
                .filter(|(_, line)| is_match(line, pattern, flags))
                .map(|(ind, line)| {
                    let line = if flags.line_numbers {
                        format!("{}:{}", ind + 1, line)
                    } else {
                        line
                    };

                    if flags.only_names {
                        return file_name.to_string();
                    };

                    if files.len() > 1 {
                        format!("{}:{}", file_name, line)
                    } else {
                        line
                    }
                })
                .collect::<Vec<String>>())
        })
        .filter(|d| match d.as_deref() {
            Ok([]) => false,
            _ => true,
        })
        .collect();

    let mut sources: Vec<_> = sources?.into_iter().flatten().collect();
    sources.dedup();

    Ok(sources.into_iter().collect())
}

File: elixir/lib/medium/grep.ex
defmodule Grep do
  @line_numbers "-n"
  @only_file_names "-l"
  @case_insensitive "-i"
  @exact_match "-x"
  @inverse_match "-v"

  @spec grep(String.t(), [String.t()], [String.t()]) :: String.t()
  def grep(pattern, flags, files) do
    Enum.map(files, fn file -> grep_file(file, pattern, flags, length(files) > 1) end)
    |> List.flatten()
    |> Enum.join("")
  end

  defp grep_file(file, pattern, flags, multiple_files?) do
    file
    |> File.stream!()
    |> Stream.with_index()
    |> Stream.filter(fn {line, _ind} -> line_matches?(flags, line, pattern) end)
    |> maybe_add_line_numbers(@line_numbers in flags and @only_file_names not in flags)
    |> maybe_prepend_with_file_name(file, multiple_files? and @only_file_names not in flags)
    |> maybe_replace_matches_with_file_name(file, @only_file_names in flags)
  end

  defp maybe_replace_matches_with_file_name([], _file, _), do: []
  defp maybe_replace_matches_with_file_name(_lines, file, true), do: ["#{file}\n"]
  defp maybe_replace_matches_with_file_name(lines, _file, false), do: lines

  defp maybe_add_line_numbers(lines, true) do
    Enum.map(lines, fn {line, ind} -> "#{ind + 1}:#{line}" end)
  end

  defp maybe_add_line_numbers(lines, false) do
    Enum.map(lines, fn {line, _ind} -> line end)
  end

  defp maybe_prepend_with_file_name(lines, file, true) do
    Enum.map(lines, fn line -> "#{file}:#{line}" end)
  end

  defp maybe_prepend_with_file_name(lines, _file, false), do: lines

  defp line_matches?(flags, line, pattern) do
    {line, pattern} =
      cond do
        @case_insensitive in flags -> {String.downcase(line), String.downcase(pattern)}
        true -> {line, pattern}
      end

    matches? =
      cond do
        @exact_match in flags -> String.trim(line) == pattern
        true -> String.contains?(line, pattern)
      end

    cond do
      @inverse_match in flags -> !matches?
      true -> matches?
    end
  end
end
